cmd/tlin/main.go

```diff
@@ -276,15 +279,54 @@ func printIssues(logger *zap.Logger, issues []tt.Issue) {
 	}
 	sort.Strings(sortedFiles)
 
-	for _, filename := range sortedFiles {
-		fileIssues := issuesByFile[filename]
-		sourceCode, err := internal.ReadSourceCode(filename)
+	if jsonOutput == "" {
+		for _, filename := range sortedFiles {
+			fileIssues := issuesByFile[filename]
+			sourceCode, err := internal.ReadSourceCode(filename)
+			if err != nil {
+				logger.Error("Error reading source file", zap.String("file", filename), zap.Error(err))
+				continue
+			}
+			output := formatter.GenetateFormattedIssue(fileIssues, sourceCode)
+			fmt.Println(output)
+		}
+	} else {
+		issuesEssentialsByFile := make(map[string][]tt.IssueEssential)
+		for _, filename := range sortedFiles {
+			fileIssues := issuesByFile[filename]
+			sourceCode, err := internal.ReadSourceCode(filename)
+			if err != nil {
+				logger.Error("Error reading source file", zap.String("file", filename), zap.Error(err))
+				continue
+			}
+
+			for _, issue := range fileIssues {
+				codeSnippet := formatter.GetCodeSnippet(issue, sourceCode)
+				issueEssential := tt.IssueEssential{
+					Rule:       issue.Rule,
+					Category:   issue.Category,
+					Message:    issue.Message,
+					Suggestion: issue.Suggestion,
+					Note:       issue.Note,
+					Snippet:    codeSnippet,
+					Confidence: issue.Confidence,
+				}
+				issuesEssentialsByFile[filename] = append(issuesEssentialsByFile[filename], issueEssential)
+			}
+		}
+
+		d, err := json.Marshal(issuesEssentialsByFile)
 		if err != nil {
-			logger.Error("Error reading source file", zap.String("file", filename), zap.Error(err))
-			continue
+			logger.Error("Error marshalling issues to JSON", zap.Error(err))
+			return
+		}
+		f, err := os.Create(jsonOutput)
+		if err != nil {
+			logger.Error("Error creating JSON output file", zap.Error(err))
+			return
 		}
-		output := formatter.GenetateFormattedIssue(fileIssues, sourceCode)
-		fmt.Println(output)
+		defer f.Close()
+		f.Write(d)
 	}
```

Comment:
If you apply the suggestion that I mentioned below, this else block can be modified as follows:

```go
// else block start here
d, err := json.Marshal(issuesByFile)
if err != nil {
	logger.Error("Error marshalling issues to JSON", zap.Error(err))
	return
}
f, err := os.Create(jsonOutput)
if err != nil {
	logger.Error("Error creating JSON output file", zap.Error(err))
	return
}
defer f.Close()
_, err = f.Write(d)
if err != nil {
	logger.Error("Error writing JSON output file", zap.Error(err))
	return
}

// ...
```
Since the encoding is processed directly from the `Issue` type, we can reduce unnecessary proccessing here.